portfolio/src/main/java/com/google/sps/TimeRange.java

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps;
+
+import java.util.Comparator;
+
+/**
+ * Class representing a span of time, enforcing properties (e.g. start comes before end) and
+ * providing methods to make ranges easier to work with (e.g. {@code overlaps}).
+ */
+public final class TimeRange {
+  public static final int START_OF_DAY = getTimeInMinutes(0, 0);
+  public static final int END_OF_DAY = getTimeInMinutes(23, 59);
+
+  public static final TimeRange WHOLE_DAY = new TimeRange(0, 24 * 60);
+
+  /**
+   * A comparator for sorting ranges by their start time in ascending order.
+   */
+  public static final Comparator<TimeRange> ORDER_BY_START = new Comparator<TimeRange>() {
+    @Override
+    public int compare(TimeRange a, TimeRange b) {
+      return Long.compare(a.start, b.start);
+    }
+  };
+
+  /**
+   * A comparator for sorting ranges by their end time in ascending order.
+   */
+  public static final Comparator<TimeRange> ORDER_BY_END = new Comparator<TimeRange>() {
+    @Override
+    public int compare(TimeRange a, TimeRange b) {
+      return Long.compare(a.end(), b.end());
+    }
+  };
+
+  private final int start;
+  private final int duration;
+
+  private TimeRange(int start, int duration) {
+    this.start = start;
+    this.duration = duration;
```

Comment:
I noticed further below you do a check to see if duration is less than or equal to 0. That raises an interesting question - are negative durations are allowed? The logic below seems based on the assumption that the start of a time range is chronologically before the end of the time range. If so then it's worth enforcing this somehow to prevent users of this class from shooting themselves in the foot. One easy thing to do would be to throw an IllegalArgumentException in the constructor if a negative duration is passed in. Most languages have a native type for nonnegative integers, but I don't think Java does.